### ISSUE TITLE
asynchronous providers

### ISSUE THREAD

**Issue by rehfeldchris** [angular/angular#23279]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[ X] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior

Angular doesn't seem to support providers which asynchronously create their object / service. For those who want to load config from a server dynamically, they seem to be encouraged to use a kludge comprised of creating a `ConfigService` and then using `APP_INITIALIZER` to delay the bootstrapping just long enough to make an async call and inject the results into the `ConfigService`. But, this is a hack, and is buggy with race conditions - the `ConfigService` may potentially be injected into other services, and be used / read from before this bootstrapping happens, causing bugs because the object gets used before it's ready.

Here's a plunk which demonstrates the problem: https://plnkr.co/edit/plWAT5i5BFtjastSdaVy?p=preview

Notice the problem is that another service will read from the `ConfigService` before it's truly finished being "constructed" IMO, since we need to construct the object, and then finish initializing it's state the `APP_INITIALIZER` step. I think reading config in a constructor is pretty common programming practice.

## Expected behavior

I think it would be great to be able to do something like:

```
export function asyncConfigServiceFactory(api: ApiService) {
    return new Promise((resolve, reject) => {
        api.get('my-config.json').then(response => {
            resolve(new ConfigService(response));
        });
    });
}

@NgModule({
  providers: [
    ApiService,
    {
      provide: ConfigService,
      useAsyncFactory: asyncConfigServiceFactory,
      deps: [ApiService]
    },
  ],
})
```

I think that's a pretty straight forward way to do it, but whatever you think is the best way to address it.

## What is the motivation / use case for changing the behavior?

My use case is to fetch config from a server, and stick it into a service, and then inject this `ConfigService` into other services. I do not want to inject `ConfigService` into other stuff until it's fully constructed and ready to be used - in other words, until the asynchronous request to the server has completed and my `ConfigService` has all the values populated into it.

I've seen quite a few other users trying to accomplish the same thing on Stackoverflow.

## Environment

<pre><code>
Angular version: 5.1

Others:

</code></pre>

Other issues that would be well served by async providers:
https://github.com/angular/angular/issues/19448  
https://github.com/angular/angular/issues/20383

**Comment by mischkl** [comment]
Having exactly this problem currently! ConfigService initialized with APP_INITIALIZER, but it's too late because some services are bootstrapped beforehand. :(

**Comment by mlc-mlapis** [comment]
@mischkl ... but isn't it just a different organizational point of view ... because:
* if you put the code to just one ConfigService, it would work
* or because APP_INITIALIZER is a multi provider ... you can have more services ... and just their declaration order in `providers` section of the main app module decides which is initialized first.

**Comment by mischkl** [comment]
The problem is that some services are defined in other modules and imported into the app module, in which case they are initialized first.

**Comment by mlc-mlapis** [comment]
@mischkl ... hmmm, I don't think so ... `APP_INITIALIZER` tokens as `multi` providers and their dependencies are initialized first. All the rest is waiting. The problem above in the Plunker was `deps: [ConfigService, ApiService]` ... and the fact that `ApiService` is also the dependency for `ConfigService`.

**Comment by trotyl** [comment]
FYI, there're indeed asynchronous DI support in the very alpha: https://github.com/angular/angular/issues/1623 https://github.com/angular/angular/issues/2813

But have already been dropped for years.

**Comment by mischkl** [comment]
@mlc-mlapis that isn't the problem with my project AFAIK. In my case I am using Ngrx and the Effects in a feature module depend on services which in turn try to retrieve the backend URL from the ConfigService in the constructor. On the other hand, if I were to inject the service into a component ("classic" Angular as opposed to redux) it wouldn't be a problem because components are initialized after APP_INITIALIZER.

In my experience it is particularly a problem when using Ngrx Effects. As the plunkr demonstrates there may be other ways to reproduce the problem, but at least when using Effects I haven't found a way to solve the problem via structural changes.

There are workarounds, of course. One is to dynamically grab the values on every service method call. Another is to expose a promise on the ConfigService and let the API services use a .then() in the constructor. The main complaint for me is simply that APP_INITIALIZER is only a partial solution since it only holds up component initialization but doesn't know anything about the service side of things. The obvious solution would be to introduce async provider loading.

**Comment by joeldavuk** [comment]
There are structural workarounds but this does seem like a basic feature I would agree with @rehfeldchris that there isn't a "right way" to do this at present and this causes a lot of a confusion.

**Comment by zlepper** [comment]
Another way to work around this, at least until angular actually support async providers is to inject a value object, which then gets updated using Object.assign(valueObj, serverResponse). 

It's still hacky, but should at least allow other client services to use a more simple object. Just beware, this doesn't actually solve the problem of ensuring the configs are ready. It just makes the semantics of getting them a bit more elegant.

**Comment by rehfeldchris** [comment]
For those that don't mind slowing down the bootstrapping process a tiny bit extra, you can do the async requests before bootstrapping. It's messy, but it's a reliable option until we get something better:

```
function bootstrapFailed(val) {
    document.getElementById('bootstrap-fail').style.display = 'block';
    console.error('bootstrap-fail', val);
}

fetch('config.json')
    .then(response => response.json())
    .then(config => {
        if (!config || !config['isConfig']) {
            bootstrapFailed(config);
            return;
        }

        // Store the response somewhere that your ConfigService can read it.
        window['tempConfigStorage'] = config;

        platformBrowserDynamic()
            .bootstrapModule(AppModule)
            .catch(bootstrapFailed);
    })
    .catch(bootstrapFailed);
```

and then use a more normal service creation strategy

```
export function configServiceFactory() {
    return new ConfigService(window['tempConfigStorage']);
}

@NgModule({
    providers: [{provide: ConfigService, useFactory: configServiceFactory}],
    bootstrap: [AppComponent]
})
export class AppModule {}
```

**Comment by jimbarrett33** [comment]
A solution to this would also promote **real** re-usable Angular libraries. We have an Angular 6 Library that is just domain services that I want others in the org to use. Currently, configured like `MyModule.forRoot( environment.whateverConfig )` in app.module.ts. The issue is some of the settings in `environment.whateverConfig` need to be provided by a call to the server for which we use `APP_INITIALIZER`. The problem is forRoot() is executed before the promise in the APP_INITIALIZER provider finishes, so the library does not get the correct config.

If someone on the Angular team could comment on the priority of this item and if/when it is on their radar to implement, that would be great.

**Comment by gagle** [comment]
This feature would be so useful for implementing reusable services in a more natural way only by injecting object-value tokens.

I also have one npm package with domain services. These services receive the base url for all the endpoints with a token. This irl comes from a fetch to the server. Currently the only way to achieve this is by fetching the url before the angular boot and then creating an extra provider that I pass to the bootstrap function instead of using a global variable.

**Comment by gogakoreli** [comment]
I have the same problem. I am trying to provide ConfigService using APP_INITIALIZER and use ConfigService to provide InjectionToken via useFactory, which has dependency on ConfigService. The problem is that inside the useFactory I get the ConfigService but it is not resolved yet. ConfigService should have config object ready to use, which is not true. By the way console.log inside useFactory is executed first and then comes the console.log from inside the APP_INITIALIZER.

**Comment by mdarefull** [comment]
+1 I wanted to add my to support this feature request. Angular has grown a lot since its first release and the entire community has grown as well, patterns has been creating and a lot of idea has been adopted from other successful frameworks as well.
Currently, it is rare to find a project that hasn't adopted Reactive Programming or even better, Redux using ngrx. Most of the modern approaches are around Observables and hence it makes sense to also update the way we create our providers.

Specifically to my case, we are using a 3rd party library with services that expect to receive configuration data as Injection Tokens. We are storing all of our data, particularly configuration on our store and the way to access it is through selectors which returns observable.
I can only imagine that with time they will appear many more use cases because observables are everywhere so, to me, it makes perfect sense to make angular provide built-in support for these scenarios.

**Comment by MatthewDavidCampbell** [comment]
Also +1 for this.  Suggestion from @rehfeldchris is good but lacks support for static `forRoot`.  Using dynamic imports is helpful (in `main.js`):

```
const grabStuff = fetch('some/configuration')
  .then(x => pull json | text | whatever)
  .then(x => parse + save to window | storage | etc);

grabStuff
  .then(_ => import('./path/to/app/module'))
  .then(({AppModule}) => {
    platformBrowserDynamic().bootstrapModule(AppModule);
  });
```

then in the AppModule:

```
@NgModule({
  ...
  SomeModule.forRoot(get storage['stuff'])
})
export class AppModule { }
```

**Comment by MatthewDavidCampbell** [comment]
Dealing with JIT is doable even for `forRoot` statics.  Just add dynamic imports to do the AppModule import after grabbing stuff from the server [*main.js*]:

```
fetch('some url')
  .then(x => save config stuff to window | storage | etc)
  .then(_ => import('./path/to/app.module'))
  .then(({AppModule}) => {
    platformBrowserDynamic().bootstrapModule(AppModule);
  });
```

Great for JIT.  Not so great with AOT.  Would love ideas on AOT.  Maybe a dynamic export that the AOT plugin can handle?

**Comment by WhitWaldo** [comment]
Just wanted to voice my +1 for this as well. I'm not presently using Angular, but am using its injection functionality via inject-js. Ideally I'd like to be able to return an observable from my factory provider to register that I can subsequently inject.

**Comment by anymos** [comment]
Same here +1, this is deeply needed especially when you have a complex app growing ...

**Comment by anymos** [comment]
Use case :

A lib provides authentification, which is configured by a provider as : 
UserModule.forRoot({fb: {
                    appId: 'fbkeyid-dynamicvalue'
});

in a multi-tenant application, the fbkeyid-dynamicvalue should be retrieved server side, asynchronously.

Happy to have a solution on this, unless that I miss it, there is none ...

Of course, considering without modifying the UserModule lib ...

Pretty standard use case. By the way, was it not possible in angular 1.X ?

**Comment by jcimoch** [comment]
Is it going to be developed after ivy is ready?

**Comment by Tahiche** [comment]
This is extremely frustating. I beleive my case is similar to @anymos 
I´m using [angular-oauth2-oidc](https://github.com/manfredsteyer/angular-oauth2-oidc) and it expects a moduleConfig object in forRoot(), which it then exposes as a provider.
After too many hours i´m frustratingly giving up.

This is the relevant part in my forRoot() providers[]

```
{ provide: OAuthModuleConfig,
         useFactory: (service: ConfigService) => {
             service.loadConfig().then(
              (res) => {
                console.log('ervice.loadConfig res', res);
                let loadedConfig = res as OAuthModuleConfig;
                return loadedConfig;
              }
            );
        },
          deps: [ConfigService]
        },
```
so "loadedConfig" is returned as a ZoneAwarePromise of sorts, there´s no way to assign the actual value. It´s there, but you can´t assign it. It´s async but I can´t use the result... 

I see no other option but to use a window object (script src="appsettings.js") , because that WILL be available and it´s the only way i can think of keeping the config separate without too much hassle.

I hope they provide a solution for this.

**Comment by chiranjeevi-puvvula-ascendlearning-com** [comment]
+ 1

@jimbarrett33 , I'm having the same issue . Did you find a temporary solution to the problem. Before I go with the workaround of fetching data during bootstap I want to check with you.

**Comment by jimbarrett33** [comment]
@chiranjeevi-puvvula-ascendlearning-com 

Still no other solution other than using fetch. The fetch solution is not that bad unless it's a huge configuration and even then...

1) it's likely that much of the stuff you're trying to configure can be baked Angular environment files so it's not needed to configure dynamically
2) generally, not everything has to be configured support the initial loading of the app. So, for example, just use dynamic config for what's needed to show the user the first first view and put the rest on a timer or something. 

I ended up fetching only dynamic config required for services (mainly endpoint config stuff) and for those services, which had already been loaded in forRoot(), called an "updateConfig()" method on them in the "initialize" service after the fetch.

**Comment by nerdDan** [comment]
This feature is necessary.

NestJS (which is a node.js backend framework similar to Angular) supports asynchronous providers[1], and it is a highly praised feature.

[1] https://docs.nestjs.com/fundamentals/async-providers

**Comment by havran** [comment]
Can we combine fetch approach with SSR and TransferState (for loading configuration only once if we use SSR)? Any example?

**Comment by jimbarrett33** [comment]
It's been 15 months. We're now at Angular 8 with so many new features but none to solve a common problem **or even a mention that it will be, or will not be, _ever_ addressed by the Angular team**. I would take a solution to this common problem over SSR or so many other of the other _extra's_ being put into the framework.

Ngrx Effects magnifies the problem even more. If you're injecting a service that has an API that needs async configuration (endpoint host, version etc.) into an Ngrx Effect, the Effect gets an incomplete configuration because the API service is injected too soon. All calls to that service now break (because the host is null/empty). There are some workarounds proposed by the Ngrx team with mixed results/reviews.

Why do we need workarounds?

<pre>
app loads
 => app needs services for data and loads the services
    => services need to be configured, many times dynamically 
       => app makes an async call with config data and configures services
          => when the services are ready let other services and components use them.</pre>

Seems like a fairly high % use case.

**Comment by yglin** [comment]
+1 for this issue.
The case here is for [Angular Google Maps](https://angular-maps.com/), which need Google Maps API key for its module forRoot() configuration. And I would really like to store those keys on some remote server instead to maintain a text environment file, forward appreciate~

**Comment by mohyeid** [comment]
This is absolutely needed. I am struggling with this now. I only can figure out my token injection value on the APP_INITIALIZER function, but no way to assign it at this time.

**Comment by csvn** [comment]
We also need something like this. We use Auth0 for user authentication. Our different environments use different domains to separate user databases. We don't want to re-build the Angular app for every environment (3+) and loading configuration before app bootstrap right now is a bit messy.

**Comment by ebauman** [comment]
+1 this is killing me in a project where I need to inject runtime configurations. Specifically, I am using Auth0's angular2-jwt product. I need to define a whitelist at runtime based on env vars.

**Comment by LexLuengas** [comment]
As @marikt pointed out in this [StackOverflow answer](https://stackoverflow.com/a/42541778/1406725), you can pass providers to the `platformBrowserDynamic` function to avoid the global property hack:

```typescript
// Load configuration that is external from build
fetch('/config.json').then(async res => {
  const configuration = await res.json();

  platformBrowserDynamic([
    { provide: AppConfiguration, useValue: configuration },
  ])
    .bootstrapModule(AppModule)
    .catch(err => console.error(err));
});

// where `AppConfiguarion` could be any injection token as usual
class AppConfiguration {
  // ...
}
```

Afterwards `AppConfiguarion` can be injected anywhere within the bootstrapped app module.

**Comment by maplion** [comment]
+1 for this issue.

**Comment by adampaquette** [comment]
The useFactory must handle async. So we have to make every dependant async?

```typescript
provide: APP_INITIALIZER,
useFactory: (settingsHttpService: SettingsHttpService) => {
    return async () => await settingsHttpService.initialize();        
},
deps: [SettingsHttpService],
multi: true
```

Here is a list of functions to prefix with async in @angular `core.js` : 

- _callFactory
- _createProviderInstance
- Injector_.get
- createClass
- _createClass
- createPipeInstance
- createDirectiveInstance
- NgModuleRef_.get
- createViewNodes
- callViewAction
- createEmbeddedView
- createRootView
- execComponentViewsAction
- execEmbeddedViewsAction
- TemplateRef_.createEmbeddedView

Who can give guidelines?

**Comment by anymos** [comment]
Hi there, a real solution without hacking the bootstrap process, would probably a LOT better

**Comment by BruneXX** [comment]
Hi all, any suggestion to use the APP_INITIALIZER approach with universal? it seems the data is not available using univeral at all, and I see no way to use transfer state API at this point. Any suggestions will be appreciated, if it's necessary I'll report a new issue to review this.

**Comment by DmitryEfimenko** [comment]
I think making all these methods async, as suggested by @adampaquette above, would not be a good solution since it would introduce too big of a breaking change.

I think we could even be able to make this work without the new `useAsyncFactory`.
I'm not very familiar with the source code, but I think one of the ways to approach implementation would be:

* loop over all providers, identify and store these that are async.
* identify all other providers that depend on async providers and keep track of them separately since they'll have to be hydrated after the async providers are resolved.
* during bootstrapping of the module start resolving async providers
* once async providers are resolved, hydrate providers that depend on async providers
* wait for all that to happen before bootstrapping components similar to how it's done for `APP_INITIALIZER`.

If I'm not mistaken (what are the chances of that?), with this order of events `injector.get` would still be synchronous since all async providers would be resolved before the injector would be used anywhere in the application.

There could be a need to call something like `await injector.resolveAsyncProviders()` in the spec files before attempting to use TestBed.get().

I'm sure I've missed a lot of important details. Heck, the whole idea is probably dead wrong. I'd be curious to see what an Angular Core team member would have to say about that.

**Comment by ljjumper** [comment]
+1 for this issue

**Comment by KostyaTretyak** [comment]
Since the idea of asynchronous DI was previously rejected by the Angular team, perhaps the best solution here might be a [resolver][1] but for init an application.

[1]: https://angular.io/api/router/Resolve

**Comment by rodrifmed** [comment]
Why angular team rejected that?!

**Comment by gmaughan** [comment]
The `fetch` approach mentioned by @LexLuengas works fine as a workaround for browser applications. But does not work when deploying in a WebView such as with Cordova or Capacitor: "fetch is not defined".

Currently attempting to implement an XMLHttpRequest approach, but support for async provider factories would be perfect.

**Comment by mybluedog24** [comment]
@gmaughan I just faced the same issue a few days ago and ended up follow Matt Tester's answer here: https://stackoverflow.com/questions/43193049/app-settings-the-angular-way

**Comment by gmaughan** [comment]
@mybluedog24 Matt's answer doesn't solve this for me. I've found that providers can be instantiated before the APP_INITIALIZER is run. In particular, when using `@angular/fire` it seems attempt to initialise before APP_INITIALIZER is run and my config is undefined.

I've posted my XMLHttpRequest solution to the [StackOverflow question you linked](https://stackoverflow.com/a/62057120/473881).

This works for browsers and Android Web View. I'm yet to test on iOS.

**Comment by johnc-ftl** [comment]
I left a comment on the [MODULE_INITIALIZER feature request](https://github.com/angular/angular/issues/17606) that the functionality people want in a hypothetical `MODULE_INITIALIZER` could be met by async factory functions. 

Also, while some of the use-cases described above can be handled by using an `APP_INITIALIZER` (documented nicely in the article [How to Handle Async Providers in Angular`](https://medium.com/better-programming/how-to-handle-async-providers-in-angular-51884647366) ), there's a major deficiency in using APP_INITIALIZER to block while async initialization completes: It only works for initializing objects in the application root. If lazy loaded modules need async initialization, they'll have to figure out another approach.

**Comment by MeMeMax** [comment]
We need this feature too. Is there a timeplan when to implement this?

**Comment by Cassidie** [comment]
Just ran into this issue today. I cannot believe this has been an issue for so long! I feel like I have to pick which hack I want to implement to get this to work. :(

**Comment by Cassidie** [comment]
> Just ran into this issue today. I cannot believe this has been an issue for so long! I feel like I have to pick which hack I want to implement to get this to work. :(

I went with what I thought was the cleanest approach. I hope this helps someone! 

I have an oauth-config.ts:
```
export const oauthConfig: OAuthModuleConfig = {
  resourceServer: {
    sendAccessToken: true,
    allowedUrls: []
  }
};
```

And then in my auth.module.ts I have:
```
import { OAuthModule } from 'angular-oauth2-oidc';
import { oauthConfig } from './oauth-config';

@NgModule({
  imports: [
    OAuthModule.forRoot(oauthConfig)
  ],
  providers:[
    {
      provide: APP_INITIALIZER,
      multi: true,
      deps: [ConfigService],
      useFactory: (configService: ConfigService) => {
        return () => {
          //Make sure to return a promise!
          return configService.loadConfig().then(() => {
           oauthConfig.resourceServer.allowedUrls = [configService.apiUrl];
          });
        };
      }
    },
    ConfigService
]

[...]
```

**Comment by projectX21** [comment]
Can somebody assist me in the following case:
In my project the user's language will be determined in a REST-call. The Locale, which is returned in the response, should be set for the LOCALE_ID, so that the pipes (date, currency) are localized.

**Comment by BruneXX** [comment]
Hi @projectX21 I think we’ll need more information to assist you, are you using the APP_INITIALIZER to get the LOCALE data and set un your project? Where are you saving that information? Are you receiving an error or something when you try to access the data?

**Comment by projectX21** [comment]
Hi @BruneXX: thank you for your imediate answer. I've got it to work in the meantime.
`fetch(<REST URL>).then(async res => {
  const localeData = new LocaleConfiguration(await res.json());

  platformBrowserDynamic([
    { provide: LocaleConfiguration, useValue: localeData }
  ]).bootstrapModule(AppModule)
    .catch(err => console.error(err));
})

export class LocaleConfiguration {
  user: String;
  locale: String;

  constructor(localeConfig: LocaleConfig) {
    this.user = localeConfig.user.name;
    this.locale = localeConfig.user.locale;
  }
}

export interface LocaleConfig {
  user: {
    name: String;
    locale: String;
  };
}`

And in the app:
`providers: [{provide: LOCALE_ID, deps: [LanguageService], useFactory: (langService: LanguageService) => langService.getLang() }],`

LangService:
`import { Injectable } from '@angular/core';
import { Observable, of } from 'rxjs';
import { delay } from 'rxjs/operators';
import { LocaleConfiguration } from './locale-configuration';

@Injectable({
  providedIn: 'root'
})
export class LanguageService {

  constructor(private localeConfig: LocaleConfiguration) { }

  getLang(): String {
    return this.localeConfig.locale;
  }
}
`

**Comment by projectX21** [comment]
Hi @BruneXX: thank you for your imediate answer. I've got it to work in the meantime.
```
fetch(<REST URL>).then(async res => {
  const localeData = new LocaleConfiguration(await res.json());

  platformBrowserDynamic([
    { provide: LocaleConfiguration, useValue: localeData }
  ]).bootstrapModule(AppModule)
    .catch(err => console.error(err));
})

export class LocaleConfiguration {
  user: String;
  locale: String;

  constructor(localeConfig: LocaleConfig) {
    this.user = localeConfig.user.name;
    this.locale = localeConfig.user.locale;
  }
}

export interface LocaleConfig {
  user: {
    name: String;
    locale: String;
  };
}
```

in the App Module:
```
providers: [{provide: LOCALE_ID, deps: [LanguageService], useFactory: (langService: LanguageService) => langService.getLang() }],
```

LangService:
```
import { Injectable } from '@angular/core';
import { Observable, of } from 'rxjs';
import { delay } from 'rxjs/operators';
import { LocaleConfiguration } from './locale-configuration';

@Injectable({
  providedIn: 'root'
})
export class LanguageService {

  constructor(private localeConfig: LocaleConfiguration) { }

  getLang(): String {
    return this.localeConfig.locale;
  }
}
```

**Comment by wojtek-viirtue** [comment]
+1 on this, being able to create objects that can't be created until runtime that are asynchronous in nature would be a huge help.

**Comment by muuvmuuv** [comment]
Highly request from our company as well, we are using Apollo with persistent caching, which uses async local storage functionality to hydrate the cache, without this we are not able to load the cache before services make requests to our API.

Read: 
- https://apollo-angular.com/docs/recipes/automatic-persisted-queries
- https://github.com/apollographql/apollo-cache-persist#web

**Comment by cerireyhan** [comment]
it's been 3 years, maybe we should give up? 😔

**Comment by maplion** [comment]
The need hasn't diminished in 3 years; why would we give up now @cerireyhan ?

**Comment by adampaquette** [comment]
> The need hasn't diminished in 3 years; why would we give up now @cerireyhan ?

I gave up Angular because Web Assembly Blazor is way better! https://dotnet.microsoft.com/apps/aspnet/web-apps/blazor

**Comment by maplion** [comment]
@Cassidie How has this approach been working out for you?  Have you run into any race conditions at module load time?

**Comment by Cassidie** [comment]
> @Cassidie How has this approach been working out for you? Have you run into any race conditions at module load time?

I haven't ran into any issues with the approach. It seems to be pretty solid. It's unfortunate that this hasn't been fixed, but at least we've got some ideas that work!

**Comment by rafagsiqueira** [comment]
@muuvmuuv , we have been using this approach for the same purpose you mentioned:

```
import { NgModule, ModuleWithProviders, APP_INITIALIZER, InjectionToken } from '@angular/core';
import { CommonModule } from '@angular/common';
import { APOLLO_OPTIONS, Apollo } from 'apollo-angular';
import { InMemoryCache, NormalizedCacheObject, split, from, concat } from '@apollo/client/core';
import { HttpLink } from 'apollo-angular/http';
import { WebSocketLink } from '@apollo/client/link/ws';
import { setContext } from '@apollo/client/link/context';
import { persistCache } from 'apollo-cache-persist';
import { AuthService } from '../auth/auth.service';

import { initialize, resolvers } from './local.state';
import { Router } from '@angular/router';
import { PersistedData, PersistentStorage } from 'apollo-cache-persist/types';
import { getMainDefinition } from '@apollo/client/utilities';
import { OperationDefinitionNode } from 'graphql';

const API_URL = new InjectionToken<string>('ApiUrl');
const WS_URL = new InjectionToken<string>('WsUrl');
const ENV = new InjectionToken<string>('Env');

const cache = new InMemoryCache() as any;

export function persistCacheFactory() {
  const promise = new Promise<void>(function(resolve) {
    persistCache({
      cache,
      storage: window.localStorage as PersistentStorage<PersistedData<NormalizedCacheObject>>,
    }).then(function() {
      initialize(cache);
      resolve();
    });
  });
  return function() {return promise; };
}

export function provideApollo(_, uri: string, wsuri: string, env: string, auth: AuthService, router: Router, httpLink: HttpLink) {

  let token = '';

  const authMiddleware = setContext(async function(parent: any, { headers }) {
    token = auth.accessToken;
    // Grab token if there is one in storage or hasn't expired
    const exp = auth.accessTokenExpiration;

    if (exp && exp <= new Date().getTime()) {
      await auth.refresh();
    }

    return {
      headers: {
        // eslint-disable-next-line @typescript-eslint/naming-convention
        Authorization: `Bearer ${token}`,
      }
    };
  });

  const ws = new WebSocketLink({
    uri: wsuri,
    options: {
      reconnect: true,
      connectionParams: {
        authToken: token,
      },
    },
  });

  const http = httpLink.create({uri});

  const splitLink = split(
    // split based on operation type
    ({query}) => {
      const {kind, operation}: OperationDefinitionNode = getMainDefinition(query) as OperationDefinitionNode;
      return (
        kind === 'OperationDefinition' && operation === 'subscription'
      );
    },
    ws,
    http,
  );

  const link = concat(
    from([authMiddleware]), splitLink);

  return {
    link,
    cache,
    resolvers,
    connectToDevTools: env === 'local' ? true : false
  };
}

@NgModule({
  declarations: [],
  imports: [
    CommonModule
  ],
  providers: [
    Apollo
  ],
  exports: []
})
export class ApiModule {
  static forRoot(uri: string, ws: string, env: string): ModuleWithProviders<ApiModule> {
    return {
      ngModule: ApiModule,
      providers: [
        {
          provide: APP_INITIALIZER,
          useFactory: persistCacheFactory,
          multi: true
        },
        {
          provide: API_URL,
          useValue: uri
        },
        {
          provide: WS_URL,
          useValue: ws
        },
        {
          provide: ENV,
          useValue: env
        },
        {
          provide: APOLLO_OPTIONS,
          useFactory: provideApollo,
          deps: [APP_INITIALIZER, API_URL, WS_URL, ENV, AuthService, Router, HttpLink]
        },
      ]
    };
  }
};
```
If you are not using websockets, just ditch the splitlink part and use:
```
const link = concat(
    from([authMiddleware]), http);
```

**Comment by muuvmuuv** [comment]
Wow @rafagsiqueira, I want to kiss you rn. Thank you!!! I just gave up on this a week ago. It was so simple... man I really need to learn thinking simpler sometimes.

**Comment by gmaughan** [comment]
> something that no longer is supported by PWA applications, so I need another solution.

@maplion I've been using a solution I mentioned [above ](https://github.com/angular/angular/issues/23279#issuecomment-635096474) in PWA applications successfully for a while now.

See [this StackOverflow answer](https://stackoverflow.com/questions/43193049/app-settings-the-angular-way/62057120#62057120) for details.

It uses XMLHttpRequest rather than fetch to build the configuration before bootstrapping the Angular module.

I agree, it shouldn't be this hard.

Glenn.

**Comment by maplion** [comment]
@gmaughan This looks promising.  I'll try it out, thank you!

**Comment by maplion** [comment]
@gmaughan You're the man; your solution works exactly like I need.

**Comment by Hafnernuss** [comment]
Unfortunately, none of the above mechanisms worked for me. I need to initialize a module with configuration, and unfortunately even with @gmaughan's approach, those modules were imported before the code ran.

[This](https://stackoverflow.com/a/60811031/1437043) approach however worked flawlessly.

**Comment by coreConvention** [comment]
I would LOVE to see this feature! Would open up so many possibilities. Why is this still not a thing?

**Comment by Yeyu83** [comment]
Hi guys! Check this repo maybe could bring you an optimal solution...

https://github.com/Yeyu83/async-data-before-bootstrap

**Comment by alxhub** [comment]
Angular's DI system is designed to be synchronous, and we have no plans to make this kind of fundamental change, for a few reasons:

1. This would force the entire runtime to become painfully asynchronous.

_Rendering_ in Angular today is synchronous. When a component `<some-cmp>` is referenced in a template, that component is instantiated and rendered along with the rest of the template.

Having asynchronous services would make component rendering asynchronous, because any individual component could not be created without perhaps waiting for a dependency to resolve. The same is true for all DI APIs, such as `Injector.get()` - they would have to return a `Promise` because the resulting instantiation might be async. Also, all rendering APIs - `ViewContainerRef.createComponent` for example. Having every component/directive rendering operation go through a `Promise` would likely have a hugely negative impact on rendering performance.

2. DI is the wrong layer to address async dependencies.

If we make rendering asynchronous, not only is it bad for performance, but it limits how the application can surface that it's currently in a loading state. Components may want to render a loading indicator, play animations, and give other signals to the user that they're currently blocked. If the framework is forced to block before it can even instantiate the component, there's no opportunity to customize the behavior - the user just stares at a blank screen or (worse) a half rendered component.

Instead, asynchronous dependencies should be modeled within the values injected, not within the DI system itself. That is, you can provide `APP_CONFIG` with an `InjectionToken<Promise<...>>` and await the promise where the config is needed, or use `APP_INITIALIZER` to block the whole bootstrap process until the config is available. This can be done fairly cleanly:

```typescript
export function provideSafeAsync<T>(
  token: T | InjectionToken<T>,
  initializer: () => Promise<T>
): Provider[] {
  const container: { value?: T } = { value: undefined };
  return [
    {
      provide: APP_INITIALIZER,
      useValue: async () => {
        container.value = await initializer();
      },
      multi: true,
    },
    {
      provide: token,
      useFactory: () => {
        if (!inject(ApplicationInitStatus).done) {
          throw new Error(
            `Cannot inject ${token} until bootstrap is complete.`
          );
        }
        return container.value;
      },
    },
  ];
}
```

**Comment by mlc-mlapis** [comment]
@alxhub Thank you for your arguments and logical justifications. There are several fundamental ideas and conceptual links in them. But isn't the initial problem most developers come up with when talking about asynchronous providers that they are looking for a way to initialize some setup parameters before processing module imports?

**Comment by DaSchTour** [comment]
From the discussion maybe it would be a good idea to have a documentation on this topic. Or maybe some helper function like shown above with documentation and usage examples. As I didn't manage to do this I've build a wrapper around the bootstrap process and use side effects to store data for initializing the application. But I would like to have some process that is more angular like.

**Comment by VladBozhinovski** [comment]
We definitely need some kind of "silver bullet approach" for this kind of issues. Often we need appsettings in Angular.

**Comment by PhilippRoessner** [comment]
Is there any chance to see a usage example of @alxhub  's Answer? How can I make a HttpClient Call with this approach during startup?

**Comment by bryanrideshark** [comment]
> Angular's DI system is designed to be synchronous, and we have no plans to make this kind of fundamental change, for a few reasons:
> 
> 1. This would force the entire runtime to become painfully asynchronous.
> 
> _Rendering_ in Angular today is synchronous. When a component `<some-cmp>` is referenced in a template, that component is instantiated and rendered along with the rest of the template.
> 
> Having asynchronous services would make component rendering asynchronous, because any individual component could not be created without perhaps waiting for a dependency to resolve. The same is true for all DI APIs, such as `Injector.get()` - they would have to return a `Promise` because the resulting instantiation might be async. Also, all rendering APIs - `ViewContainerRef.createComponent` for example. Having every component/directive rendering operation go through a `Promise` would likely have a hugely negative impact on rendering performance.
> 
> 2. DI is the wrong layer to address async dependencies.
> 
> If we make rendering asynchronous, not only is it bad for performance, but it limits how the application can surface that it's currently in a loading state. Components may want to render a loading indicator, play animations, and give other signals to the user that they're currently blocked. If the framework is forced to block before it can even instantiate the component, there's no opportunity to customize the behavior - the user just stares at a blank screen or (worse) a half rendered component.
> 
> Instead, asynchronous dependencies should be modeled within the values injected, not within the DI system itself. That is, you can provide `APP_CONFIG` with an `InjectionToken<Promise<...>>` and await the promise where the config is needed, or use `APP_INITIALIZER` to block the whole bootstrap process until the config is available. This can be done fairly cleanly:
> 
> ```ts
> export function provideSafeAsync<T>(
>   token: T | InjectionToken<T>,
>   initializer: () => Promise<T>
> ): Provider[] {
>   const container: { value?: T } = { value: undefined };
>   return [
>     {
>       provide: APP_INITIALIZER,
>       useValue: async () => {
>         container.value = await initializer();
>       },
>       multi: true,
>     },
>     {
>       provide: token,
>       useFactory: () => {
>         if (!inject(ApplicationInitStatus).done) {
>           throw new Error(
>             `Cannot inject ${token} until bootstrap is complete.`
>           );
>         }
>         return container.value;
>       },
>     },
>   ];
> }
> ```

What if there was a new category of provider, which was only made available to `templates`? 

Imagine the following hypothetical:

````ts

@Component({ 
   standalone: true,
   template: `
   <div *ngIf="userIsLoggedIn">
      Only show if the user is logged in.
   </div>
   <div *ngIf="userIsAdministrator">
      Show expensive stuff that only applies if the user is an admin; Load expensive admin-only components and modules to boot.
   </div>
   `,
   templateProviders: [{
     provide: 'userIsLoggedIn',
     async: {
       defaultValue: false,
       useFactory: () => import('get-user-is-logged-in.function').then(m => m.getUserIsLoggedInFunction),
       deps: () => [import('user-service').then(m => m.UserService)]
     }
   },
   {
     provide: 'userIsAdministrator',
     async: {
       defaultValue: false,
       useFactory: () => import('get-user-is-administrator.function').then(m => m.getUserIsAdministratorFunction),
       deps: () => [
          import('expensive-and-inefficient-admin-user-service').then(m => m.AdminUserService),
          import('another-expensive-dependency-to-demonstrate').then(m => m.AnotherExpensiveDependency)
          ]
     }
   }
   ]
})
export class Foo {}
````

**Comment by szymon-obsertio** [comment]
> 2\. DI is the wrong layer to address async dependencies.

there is definitely number of cases where various workaround as in use instead of providing reasonable solution within Angular. One more example is when you try to combine `ErrorHandler` provider with some external tracking tool and any async configuration - eg I have [similar problem to this bugsnag initialisation](https://stackoverflow.com/questions/53496305/ionicangularbugsnag-custom-error-handler-and-async-appversion)  - I'd like to share tracked bugsnag session in browser extension and I need async to read the session.

I did hope that I could eg use `deps: [APP_INITIALIZER]` and expect this to be a case where I can expect the proposed `provideSafeAsync` to be already in a resolved state. This way we could avoid introducing general asynchronous DI but also provide reasonable way to just make the async `APP_INITIALIZER` helpful for quite a lot of cases.

**Comment by ciekawy** [comment]
> 2\. DI is the wrong layer to address async dependencies.

there is definitely number of cases where various workaround as in use instead of providing reasonable solution within Angular. One more example is when you try to combine `ErrorHandler` provider with some external tracking tool and any async configuration - eg I have [similar problem to this bugsnag initialisation](https://stackoverflow.com/questions/53496305/ionicangularbugsnag-custom-error-handler-and-async-appversion)  - I'd like to share tracked bugsnag session in browser extension and I need async to read the session.

I did hope that I could eg use `deps: [APP_INITIALIZER]` and expect this to be a case where I can expect the proposed `provideSafeAsync` to be already in a resolved state. This way we could avoid introducing general asynchronous DI but also provide reasonable way to just make the async `APP_INITIALIZER` helpful for quite a lot of cases.

**Comment by Lonli-Lokli** [comment]
But you actually can use app_initializer for that

**Comment by ciekawy** [comment]
@Lonli-Lokli so the case is that
1. `ErrorHandler` provider needs to return an instance even before the main here in the [application-ref](https://github.com/angular/angular/blob/07017a7bd30c6bb7bd1b94a9fd1b112ee68c9ced/packages/core/src/application_ref.ts#L216)
2. `BugsnagErrorHandler` needs the `Bugsnag` client to be already initialised
3. to create `Bugsnag` client I need to read the sessionID from `browser.storage.local` which has async access.
Of course this is quite a corner case and you can always expect the vendor should consider that.  For now I use a workaround and I'd prefer to relay on Angular support

**Comment by Lonli-Lokli** [comment]
@ciekawy Yep, but on errorHandler init line all app_initializer depa should be already created, ie async loaded.
Unfortunately it's not guaranteed with angular and the only way to make it work is to put canactivateguard with required dependency on first route to make it load.
If it's not clear, you can create a sample with notworkijlng behaviour and I will try to adopt to make it work

**Comment by ciekawy** [comment]
route is way too late. I can always do typical approach to use 
```
(async () => {
  platformBrowserDynamic([{
    provide: SESSION_ID,
    useValue: await resolveSessionId()
  }]).bootstrapModule(BackgroundModule);
})();
```

I believe it should be possible to handle async providers (provide resolved values) within `APP_INITIALIZER` as anyway returning promise/observable will block app initialisation.

**Comment by Lonli-Lokli** [comment]
Route is never late as it creating now before app component

**Comment by ciekawy** [comment]
In my case I need to resolve async value to create `ErrorHandler ` (and anyway I want errors to be intercepted from the very beginning including initialisation phase) and secondly I'm reusing some logic in browser extension background script having no routes (I know the second is pretty specific)

**Comment by ak99372** [comment]
While APP_INITIALIZER works well for app initial loading, it does not work for lazy loaded modules that also need to initialize their providers long after the app started. Scenarios where provider configuration might not be known at app start or where providers are incrementally setup/loaded later in the app's run. We're on 15th version while still thinking in terms of rigid and statically configured monoliths without async option (not even suggested workaround/example after almost 5 years)?

**Comment by smaillns** [comment]
I'm facing the same issue, there are some services initialized before providing APP_INITIALIZER, and I'm also experiencing the issue when using a library mentioned by @jimbarrett33    [link](https://github.com/angular/angular/issues/23279#issuecomment-412108311) . is there any relevant solution ? thanks in advance

**Comment by Azbesciak** [comment]
As a 300th "liking" of this issue I would ask is it going to happen?

**Comment by WhitWaldo** [comment]
> As a 300th "liking" of this issue I would ask is it going to happen?

It's been 5 years with no traction - don't hold your breath.

**Comment by csisy-bt4w** [comment]
`APP_INITIALIZER` is not even an appropriate solution. What about lazy-loaded modules? Also, the `ENVIRONMENT_INITIALIZER` does not accept async functions.

**Comment by Vaili** [comment]
I solved it by delaying the import of main module in bootstrap.ts untill the config is fetched. 

`
async function bootstrapApp() {
  await myApiReqeustThatShouldProvideTheConfiguration();

  const authModule =  await import('./app/authentication-layer/authentication-layer.module');
  
  platformBrowserDynamic()
    .bootstrapModule(authModule.AuthenticationLayerModule)
    .catch((err) => console.error(err));
}` 

The resence of import on top of the bootstrap file is starting the store even before platform browser. This is dangerous

**Comment by Char2sGu** [comment]
Enabling Injection Context in route configs’ “loadChildren” resolvers could allow you to programmatically generate child route configs with dynamic “providers”. See https://github.com/angular/angular/issues/51532#issuecomment-1955238331 for a working workaround to enable injection context for loadChildren.

**Comment by ak99372** [comment]
Using "LoadChildren" type on a route accepts async Routes but also async NgModuleFactory so you can asynchronously create/load module (including providers). For instance using async function that returns ModuleWithProviders where providers are loaded asynchronously.

```
{ 
 path: "lazy-module", 
 loadChildren: lazyLoadModuleWithProviders() 
}

export function lazyLoadModuleWithProviders(): LoadChildren {
	return async () => {
		const module = await ... //add async calls here that feed with your module + providers
		const factory = new LazyModuleFactory(module);
		return factory;
	}
}
```
**Factory Example:**
```
export class LazyModuleFactory<T> extends NgModuleFactory<T> {

	constructor(private moduleWithProviders: ModuleWithProviders<T>) {
		super();
	}

	get moduleType(): Type<T> {
		return this.moduleWithProviders.ngModule;
	}

	create(parentInjector: Injector | null): NgModuleRef<T> {
		const providers = this.moduleWithProviders.providers as StaticProvider[];
		const injector = Injector.create({
			parent: parentInjector,
			providers
		});

		return createNgModule(this.moduleType, injector);
	}
}
```

 Of course the limitation is that this only happens at the time of Module creation, which is when the providers would be materialized anyways (and still better than doing it on APP_INITIALIZER) but that's as close as you can get to async provider right now.

NOTE: The sad part of this workaround is that NgModuleFactory has "recently" been marked as deprecated :( so one step forward, one step back )

**Comment by Azbesciak** [comment]
I was investigating the multer's code, and I see that feature is available in nest js
https://github.com/nestjs/nest/blob/master/packages/platform-express/multer/multer.module.ts

Would be great to have it also in angular.

**Comment by kaiden-desperation** [comment]
I gathered the answer https://honeymonty.github.io/dapper7/

**Comment by JeanMeche** [comment]
#68248 will land an `injectAsync` helper in v22 that should address some of the usecases mentioned here.
